### PR TITLE
Move pipeline stages into subpackage

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

Tidies up the `codegen` package by moving the pipeline stages into a subpackage.

Consists almost entirely of mechanical refactoring steps. 

The exceptions are
* Moving a small number of helper methods around
* Improving the description of some stages
* Improving some code comments

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: hack/generator/pkg/codegen/pipeline/apply_kubernetes_resource_interface.go [hunk 1]
```diff
@@ -3,21 +3,22 @@
...
-// applyKubernetesResourceInterface ensures that every Resource implements the KubernetesResource interface
-func applyKubernetesResourceInterface(idFactory astmodel.IdentifierFactory) PipelineStage {
+// ApplyKubernetesResourceInterface ensures that every Resource implements the KubernetesResource interface
+func ApplyKubernetesResourceInterface(idFactory astmodel.IdentifierFactory) Stage {
 
-	return MakePipelineStage(
+	return MakeStage(
 		"applyKubernetesResourceInterface",
-		"Ensures that every resource implements the KubernetesResource interface",
+		"Add the KubernetesResource to every resource",
```
Comment (matthchr) [inline hack/generator/pkg/codegen/pipeline/apply_kubernetes_resource_interface.go:21]:
```suggestion
		"Add the KubernetesResource interface to every resource",
```

File: hack/generator/pkg/codegen/pipeline/assert_types_collection_valid.go [hunk 2]
```diff
@@ -3,21 +3,21 @@
...
+// AssertTypesCollectionValid creates a Stage that ensures that each reachable type in the types collection
 // has TypeName's that are all reachable as well. This check fails if there is any TypeName that refers to a type that doesn't
 // exist.
-func assertTypesCollectionValid() PipelineStage {
-	return MakePipelineStage(
+func AssertTypesCollectionValid() Stage {
+	return MakeStage(
 		"assertTypesStructureValid",
-		"Asserts that the types collection is valid",
+		"Verify that all named types are properly defined",
```
Comment (matthchr) [inline hack/generator/pkg/codegen/pipeline/assert_types_collection_valid.go:20]:
```suggestion
		"Verify that all local TypeNames refer to a type",
```

What you have now makes it sound like it's looking at the structure of named types when really it's looking at the references between types (named or not)

File: hack/generator/pkg/codegen/pipeline/create_arm_types.go [hunk 3]
```diff
@@ -15,12 +15,12 @@ import (
...
-// createARMTypes walks the type graph and builds new types for communicating
+// CreateARMTypes walks the type graph and builds new types for communicating
 // with ARM
-func createARMTypes(idFactory astmodel.IdentifierFactory) PipelineStage {
-	return MakePipelineStage(
+func CreateARMTypes(idFactory astmodel.IdentifierFactory) Stage {
+	return MakeStage(
 		"createArmTypes",
-		"Creates ARM types",
+		"Create custom types for interaction with ARM",
```
Comment (matthchr) [inline hack/generator/pkg/codegen/pipeline/create_arm_types.go:23]: minor:
```suggestion
		"Create types for interaction with ARM",
```

Unsure what custom is alluding to here but I don't think these types are more custom than any of our other types?

File: hack/generator/pkg/codegen/pipeline/flatten_resources.go [hunk 4]
```diff
@@ -3,18 +3,19 @@
...
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
 )
 
-// flattenResources flattens any resources directly inside other resources
-func flattenResources() PipelineStage {
-	return MakePipelineStage(
+// FlattenResources flattens any resources directly inside other resources
+func FlattenResources() Stage {
+	return MakeStage(
 		"flatten-resources",
```
Comment (matthchr) [inline hack/generator/pkg/codegen/pipeline/flatten_resources.go:19]:
```suggestion
		"flattenResources",
```

**Review by matthchr — APPROVED** [review]
Some minor wording suggestions

File: hack/generator/pkg/codegen/code_generator.go [hunk 5]
```diff
@@ -87,101 +77,101 @@ func NewCodeGeneratorFromConfig(configuration *config.Configuration, idFactory a
...
 
-		deleteGeneratedCode(configuration.FullTypesOutputPath()),
+		pipeline.DeleteGeneratedCode(configuration.FullTypesOutputPath()),
 
-		exportPackages(configuration.FullTypesOutputPath()).
+		pipeline.ExportPackages(configuration.FullTypesOutputPath()).
 			RequiresPrerequisiteStages("deleteGenerated"),
 
-		exportControllerResourceRegistrations(configuration.FullTypesRegistrationOutputFilePath()).UsedFor(ARMTarget),
+		pipeline.ExportControllerResourceRegistrations(configuration.FullTypesRegistrationOutputFilePath()).UsedFor(pipeline.ARMTarget),
```
Comment (babbageclunk) [inline hack/generator/pkg/codegen/code_generator.go:160]: Whoa there *are* a lot of stages!

File: hack/generator/pkg/codegen/pipeline/crossplane_add_at_provider.go [hunk 6]
```diff
@@ -14,10 +14,10 @@ import (
...
 
-// addCrossplaneAtProvider adds an "AtProvider" property as the sole property in every resource status
-func addCrossplaneAtProvider(idFactory astmodel.IdentifierFactory) PipelineStage {
+// AddCrossplaneAtProvider adds an "AtProvider" property as the sole property in every resource status
+func AddCrossplaneAtProvider(idFactory astmodel.IdentifierFactory) Stage {
 
-	return MakePipelineStage(
+	return MakeStage(
 		"addCrossplaneAtProviderProperty",
 		"Adds an 'AtProvider' property on every status",
```
Comment (babbageclunk) [inline hack/generator/pkg/codegen/pipeline/crossplane_add_at_provider.go:22]:
```suggestion
		"Add an 'AtProvider' property on every status",
```

File: hack/generator/pkg/codegen/pipeline/crossplane_add_embedded_resource_spec.go [hunk 7]
```diff
@@ -15,10 +15,10 @@ import (
...
 
-// addCrossplaneEmbeddedResourceSpec puts an embedded runtimev1alpha1.ResourceSpec on every spec type
-func addCrossplaneEmbeddedResourceSpec(idFactory astmodel.IdentifierFactory) PipelineStage {
+// AddCrossplaneEmbeddedResourceSpec puts an embedded runtimev1alpha1.ResourceSpec on every spec type
+func AddCrossplaneEmbeddedResourceSpec(idFactory astmodel.IdentifierFactory) Stage {
 
-	return MakePipelineStage(
+	return MakeStage(
 		"addCrossplaneEmbeddedResourceSpec",
 		"Adds an embedded runtimev1alpha1.ResourceSpec to every spec type",
```
Comment (babbageclunk) [inline hack/generator/pkg/codegen/pipeline/crossplane_add_embedded_resource_spec.go:23]:
```suggestion
		"Add an embedded runtimev1alpha1.ResourceSpec to every spec type",
```

File: hack/generator/pkg/codegen/pipeline/crossplane_add_embedded_resource_status.go [hunk 8]
```diff
@@ -13,10 +13,10 @@ import (
...
 
-// addCrossplaneEmbeddedResourceStatus puts an embedded runtimev1alpha1.ResourceStatus on every spec type
-func addCrossplaneEmbeddedResourceStatus(idFactory astmodel.IdentifierFactory) PipelineStage {
+// AddCrossplaneEmbeddedResourceStatus puts an embedded runtimev1alpha1.ResourceStatus on every spec type
+func AddCrossplaneEmbeddedResourceStatus(idFactory astmodel.IdentifierFactory) Stage {
 
-	return MakePipelineStage(
+	return MakeStage(
 		"addCrossplaneEmbeddedResourceStatus",
 		"Adds an embedded runtimev1alpha1.ResourceStatus to every status type",
```
Comment (babbageclunk) [inline hack/generator/pkg/codegen/pipeline/crossplane_add_embedded_resource_status.go:21]:
```suggestion
		"Add an embedded runtimev1alpha1.ResourceStatus to every status type",
```

File: hack/generator/pkg/codegen/pipeline/crossplane_add_for_provider.go [hunk 9]
```diff
@@ -14,11 +14,11 @@ import (
...
-// addCrossplaneForProvider adds a "ForProvider" property as the sole property in every resource spec
+// AddCrossplaneForProvider adds a "ForProvider" property as the sole property in every resource spec
 // and moves everything that was at the spec level down a level into the ForProvider type
-func addCrossplaneForProvider(idFactory astmodel.IdentifierFactory) PipelineStage {
+func AddCrossplaneForProvider(idFactory astmodel.IdentifierFactory) Stage {
 
-	return MakePipelineStage(
+	return MakeStage(
 		"addCrossplaneForProviderProperty",
 		"Adds a 'ForProvider' property on every spec",
```
Comment (babbageclunk) [inline hack/generator/pkg/codegen/pipeline/crossplane_add_for_provider.go:23]:
```suggestion
		"Add a 'ForProvider' property on every spec",
```

File: hack/generator/pkg/codegen/pipeline/crossplane_add_owner_properties.go [hunk 10]
```diff
@@ -14,10 +14,10 @@ import (
...
 
-// addCrossplaneOwnerProperties adds the 3-tuple of (xName, xNameRef, xNameSelector) for each owning resource
-func addCrossplaneOwnerProperties(idFactory astmodel.IdentifierFactory) PipelineStage {
+// AddCrossplaneOwnerProperties adds the 3-tuple of (xName, xNameRef, xNameSelector) for each owning resource
+func AddCrossplaneOwnerProperties(idFactory astmodel.IdentifierFactory) Stage {
 
-	return MakePipelineStage(
+	return MakeStage(
 		"addCrossplaneOwnerProperties",
 		"Adds the 3-tuple of (xName, xNameRef, xNameSelector) for each owning resource",
```
Comment (babbageclunk) [inline hack/generator/pkg/codegen/pipeline/crossplane_add_owner_properties.go:22]:
```suggestion
		"Add the 3-tuple of (xName, xNameRef, xNameSelector) for each owning resource",
```

File: hack/generator/pkg/codegen/pipeline/mark_storage_version.go [hunk 11]
```diff
@@ -14,9 +14,9 @@ import (
...
 )
 
-// markStorageVersion creates a PipelineStage to mark a particular version as a storage version
-func markStorageVersion() PipelineStage {
-	return MakePipelineStage(
+// MarkStorageVersion creates a Stage to mark a particular version as a storage version
+func MarkStorageVersion() Stage {
+	return MakeStage(
 		"markStorageVersion",
 		"Marking the latest version of each resource as the storage version",
```
Comment (babbageclunk) [inline hack/generator/pkg/codegen/pipeline/mark_storage_version.go:21]:
```suggestion
		"Mark the latest version of each resource as the storage version",
```

File: hack/generator/pkg/codegen/pipeline/remove_embedded_resources.go [hunk 12]
```diff
@@ -12,8 +12,8 @@ import (
 	"github.com/Azure/azure-service-operator/hack/generator/pkg/codegen/embeddedresources"
 )
 
-func removeEmbeddedResources() PipelineStage {
-	return MakePipelineStage(
+func RemoveEmbeddedResources() Stage {
+	return MakeStage(
 		"removeEmbeddedResources",
 		"Removes properties that point to embedded resources. Only removes structural aspects of embedded resources, Id/ARMId references are retained.",
```
Comment (babbageclunk) [inline hack/generator/pkg/codegen/pipeline/remove_embedded_resources.go:18]:
```suggestion
		"Remove properties that point to embedded resources. Only removes structural aspects of embedded resources, Id/ARMId references are retained.",
```

**Review by babbageclunk — APPROVED** [review]
Looks good - I commented on a few missed descriptions. 😁 
This moves them out of the codegen package but still keeps them in the same pipeline package, so maybe there's still a bit of a problem with all of the helper functions for the different stages being in the same namespace? That can be fixed progressively though.

**Pull request merged**